src/main/java/uk/gov/digital/ho/hocs/workflow/api/bpmn/BusinessEventService.java

```diff
@@ -0,0 +1,112 @@
+
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.UUID;
+
+import static uk.gov.digital.ho.hocs.workflow.api.WorkflowConstants.DATA_FIELD_UPDATED_EVENT_NAME;
+import static uk.gov.digital.ho.hocs.workflow.api.WorkflowConstants.PREVIOUS_VALUE_KEY;
+
+@Service
+@Slf4j
+public class BusinessEventService {
+    private final AuditClient auditClient;
+    private final CaseworkClient caseworkClient;
+    private final BpmnService bpmnService;
+    private final FormService formService;
+
+    public BusinessEventService(AuditClient auditClient,
+                                CaseworkClient caseworkClient,
+                                BpmnService bpmnService,
+                                FormService formService) {
+        this.auditClient = auditClient;
+        this.caseworkClient = caseworkClient;
+        this.bpmnService = bpmnService;
+        this.formService = formService;
+    }
+
+    public void createDataFieldUpdatedEvent(String caseUUIDString, String stageUUIDString, String screenName, String fieldName, String newValue) {
+        UUID caseUUID = UUID.fromString(caseUUIDString);
+        UUID stageUUID= UUID.fromString(stageUUIDString);
+
+        String previousValueKey = String.format(PREVIOUS_VALUE_KEY, fieldName);
+
+        GetCaseworkCaseDataResponse caseData = caseworkClient.getCase(caseUUID);
+        String previousValue = caseData.getData().get(previousValueKey);
+
+        if (Objects.equals(previousValue, newValue)) {
+            return;
+        }
+
+        HocsSchema hocsSchema = formService.getFormSchema(screenName);
+
+        List<HocsFormField> fields = hocsSchema.getFields();
+        Optional<HocsFormField> hocsFormField = fields.stream().filter((field) -> Objects.equals(field.getProps().get("name"), fieldName)).findFirst();
+
+        String fieldNameLabel = getFormProperty("label", hocsFormField, String.class).orElse(fieldName);
+
+        List<Map<String, String>> choices = getChoices(hocsFormField, caseData);
+
+        String newValueLabel = getLabel(newValue, choices);
+        String previousValueLabel = getLabel(previousValue, choices);
+
+
+        DataFieldUpdatedPayload dataFieldUpdatedPayload = new DataFieldUpdatedPayload(
+            DATA_FIELD_UPDATED_EVENT_NAME,
+            fieldName,
+            newValue,
+            previousValue,
+            fieldNameLabel,
+            newValueLabel,
+            previousValueLabel
+        );
+
+        auditClient.createBusinessEvent(caseUUID, stageUUID, dataFieldUpdatedPayload);
+        bpmnService.updateCaseValue(caseUUIDString, stageUUIDString, previousValueKey, newValue);
+    }
+
+    private List<Map<String, String>> getChoices(Optional<HocsFormField> hocsFormField, GetCaseworkCaseDataResponse caseData) {
+        @SuppressWarnings("unchecked") List<Object> conditionChoices = (List<Object>) getFormProperty("conditionChoices", hocsFormField, List.class).orElse(List.of());
```

Comment:
Not blocking but this feels like a semi-odd way of doing this casting. We have places that we use `TypeReference<>`, so wonder if that can be used to simplify this.

Reply:
https://github.com/UKHomeOffice/hocs-workflow/pull/856/commits/bcc5db725dcc52471e3e0ae8fe23a40deae59578

Migrated to use TypeReferences / objectMapper.